<a href="https://colab.research.google.com/github/omullaboyev/data-science/blob/main/2_1_16_Pandas_Ranking.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Data Science va Sun'iy Intellekt Praktikum

Ma'lumotlar Tahlili. Pandas kutubxonasi

2.1 Pandas kutubxonasi bilan ishlash.

2.1.16 Reytinglash (Ranking)

**1 - eng baland reyting**

In [1]:
import pandas as pd
import numpy as np

In [2]:
# internetdan CSV faylni dataframe ga yuklab olamiz
df = pd.read_csv("https://raw.githubusercontent.com/anvarnarz/praktikum_datasets/main/usa_cars.csv", index_col = 0)
df.head()

price      brand    model  ...       state country      condition
0   6300     toyota  cruiser  ...  new jersey     usa   10 days left
1   2899       ford       se  ...   tennessee     usa    6 days left
2   5350      dodge      mpv  ...     georgia     usa    2 days left
3  25000       ford     door  ...    virginia     usa  22 hours left
4  27700  chevrolet     1500  ...     florida     usa  22 hours left

[5 rows x 12 columns]

In [3]:
# mashinalarga ishlab chiqarilgan yili bo'yicha reyting beramiz
df["rating"] = df["year"].rank()

In [4]:
df

price      brand    model  ...  country      condition  rating
0      6300     toyota  cruiser  ...      usa   10 days left    70.5
1      2899       ford       se  ...      usa    6 days left   115.0
2      5350      dodge      mpv  ...      usa    2 days left  1362.0
3     25000       ford     door  ...      usa  22 hours left   336.5
4     27700  chevrolet     1500  ...      usa  22 hours left  1362.0
...     ...        ...      ...  ...      ...            ...     ...
2494   7800     nissan    versa  ...      usa    1 days left  2005.5
2495   9200     nissan    versa  ...      usa  21 hours left  1362.0
2496   9200     nissan    versa  ...      usa  21 hours left  1362.0
2497   9200     nissan    versa  ...      usa    2 days left  1362.0
2498   9200     nissan    versa  ...      usa  21 hours left  1362.0

[2499 rows x 13 columns]

Har bir mashinaning **year** bo'yicha reytingi df da shu yilda ishlab chiqarilgan mashina nechta bo'lsa shunga teng.

In [5]:
df.sort_values("year") # year ustuni qiymatlari bo'yicha tartiblaymiz

price      brand     model  ...  country      condition  rating
32    29800  chevrolet    camaro  ...      usa  22 hours left     1.0
405      25       ford      door  ...      usa  17 hours left     2.0
545       0        gmc      door  ...      usa  18 hours left     3.0
362      25       ford    pickup  ...      usa    2 days left     4.5
322       0       ford   chassis  ...      usa  19 hours left     4.5
...     ...        ...       ...  ...      ...            ...     ...
426   55600      lexus        gx  ...      usa    2 days left  2475.5
806   28300  chevrolet  colorado  ...      usa    2 days left  2475.5
2322  15000     nissan    altima  ...      usa    2 days left  2475.5
2344  20100     nissan     rogue  ...      usa  14 hours left  2475.5
1388  17000       ford    fusion  ...      usa    2 days left  2475.5

[2499 rows x 13 columns]

*   1973 yilda ishlab chiqarilgan mashina bitta bo'lgani uchun reytingi 1 ga teng (eng baland reyting)
*   1984 yilda ishlab chiqarilgan mashina ikkita bo'lgani uchun reytingi 2 ga teng
*   1993 yilda ishlab chiqarilgan mashina uchta bo'lgani uchun reytingi 3 ga teng
*   .............
*   2020 yilda ishlab chiqarilgan mashina ko'p bo'lgani uchun reytingi 2475.5 ga teng (eng past reyting)

#`method` - reytinglash metodi
min - eng kichikdan
average - o'rtacha qiymat

In [6]:
# mashinalarga narxi bo'yicha reyting beramiz
df["rating"] = df["price"].rank(method="min") # minimumdan boshlab reyting beramiz

In [7]:
df

price      brand    model  ...  country      condition  rating
0      6300     toyota  cruiser  ...      usa   10 days left   380.0
1      2899       ford       se  ...      usa    6 days left   156.0
2      5350      dodge      mpv  ...      usa    2 days left   326.0
3     25000       ford     door  ...      usa  22 hours left  1824.0
4     27700  chevrolet     1500  ...      usa  22 hours left  2004.0
...     ...        ...      ...  ...      ...            ...     ...
2494   7800     nissan    versa  ...      usa    1 days left   475.0
2495   9200     nissan    versa  ...      usa  21 hours left   567.0
2496   9200     nissan    versa  ...      usa  21 hours left   567.0
2497   9200     nissan    versa  ...      usa    2 days left   567.0
2498   9200     nissan    versa  ...      usa  21 hours left   567.0

[2499 rows x 13 columns]

In [8]:
df.sort_values("price") # price - narxi bo'yicha tartiblaymiz.

price          brand       model  ...  country      condition  rating
410       0      chevrolet        door  ...      usa    2 days left     1.0
330       0           ford        door  ...      usa  19 hours left     1.0
331       0           ford        door  ...      usa  17 hours left     1.0
339       0           ford        door  ...      usa    2 days left     1.0
496       0           ford      pickup  ...      usa  17 hours left     1.0
...     ...            ...         ...  ...      ...            ...     ...
1215  65500           ford         srw  ...      usa  21 hours left  2495.0
277   67000          dodge  challenger  ...      usa  21 hours left  2496.0
1336  70000           ford         drw  ...      usa    2 days left  2497.0
1340  74000           ford         drw  ...      usa    2 days left  2498.0
502   84900  mercedes-benz    sl-class  ...      usa    2 days left  2499.0

[2499 rows x 13 columns]

In [10]:
# mashinalarga mileage bo'yicha reyting beramiz
df["rating"] = df["mileage"].rank(method="average") # o'rtacha qiymat bo'yicha reyting beramiz

In [13]:
df.sort_values("mileage") # mileage bo'yicha tartiblaymiz.

price      brand   model  ...  country      condition  rating
545       0        gmc    door  ...      usa  18 hours left     3.5
504     100  peterbilt   truck  ...      usa  17 hours left     3.5
1619    650       ford    door  ...      usa    2 days left     3.5
309       0  chevrolet    door  ...      usa  18 hours left     3.5
1236   4200       ford    door  ...      usa    2 days left     3.5
...     ...        ...     ...  ...      ...            ...     ...
531    2000  chevrolet  pickup  ...      usa  21 hours left  2495.0
490     475  peterbilt   truck  ...      usa  17 hours left  2496.0
516       0  peterbilt   truck  ...      usa  17 hours left  2497.0
1827   3200       ford    door  ...      usa  21 hours left  2498.0
528    1025  peterbilt   truck  ...      usa  17 hours left  2499.0

[2499 rows x 13 columns]

Mashina qancha kam yurgan bo'lsa reytingi shuncha baland.